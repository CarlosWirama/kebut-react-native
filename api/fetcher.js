
export async function fetchNYTimesAPI(endpoint, params) {
  const domain = "https://api.nytimes.com/svc";
  const url = domain + endpoint;
  const querystring = Object.keys(params)
    .map(key => key + '=' + encodeURIComponent(params[key]))
    .join('&');
  let response = await fetch(`${url}?${querystring}`);
  response = await response.json();
  return response.response.docs;
}