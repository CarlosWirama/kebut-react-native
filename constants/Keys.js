export default {
  NYTimes: '7bb6006970f94dc79a5f04ba7b16e2f2',
};
