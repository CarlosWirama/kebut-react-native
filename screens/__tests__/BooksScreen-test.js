import 'react-native';
import React from 'react';
import Screen from '../BooksScreen';
import {fetchNYTimesAPI} from '../../api/fetcher';
import renderer from 'react-test-renderer';
import Keys from '../../constants/Keys';
import mockedWindowFetch from '../__mocked-data__/windowFetch_Book';

beforeEach(() => {
  window.fetch = jest.fn().mockImplementation(
    () => Promise.resolve(mockedWindowFetch)
  );
});

const sleep = (ms) => {
	return new Promise( resolve => setTimeout(resolve, ms) );
}

it('should render initial BooksScreen', async () => {
  const tree = renderer.create(<Screen />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render BooksScreen with List after 3s', async () => {
  await sleep(3000);
  const tree = renderer.create(<Screen />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should return list when fetching book', async () => {
  const searchEndpoint = "/svc/books/v3/lists.json";
  const params = {
    list: 'e-book-fiction',
    'api-key': Keys.NYTimes,
  };
  let result = await fetchNYTimesAPI(searchEndpoint, params);
  result = result.results;
  expect(Array.isArray(result)).toBe(true);
  expect(result.length > 0).toBe(true);
  expect(typeof result[0].book_details[0].title).toBe('string');
  expect(typeof result[0].book_details[0].description).toBe('string');
  expect(typeof result[0].book_details[0].author).toBe('string');
});