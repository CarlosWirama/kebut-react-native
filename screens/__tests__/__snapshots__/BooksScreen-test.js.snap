// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BooksScreen with List after 3s 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "margin": 10,
        "marginBottom": 5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginVertical": 5,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "green",
            "borderRadius": 5,
            "borderWidth": 2,
            "flex": 1,
            "opacity": 1,
            "padding": 10,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          E-Book Fiction
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "white",
            "borderWidth": 2,
            "flex": 1,
            "opacity": 1,
            "padding": 10,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Hardcover Fiction
        </Text>
      </View>
    </View>
  </View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`should render initial BooksScreen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "margin": 10,
        "marginBottom": 5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginVertical": 5,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "green",
            "borderRadius": 5,
            "borderWidth": 2,
            "flex": 1,
            "opacity": 1,
            "padding": 10,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          E-Book Fiction
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "white",
            "borderWidth": 2,
            "flex": 1,
            "opacity": 1,
            "padding": 10,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Hardcover Fiction
        </Text>
      </View>
    </View>
  </View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
